solr/contrib/ltr/src/test/org/apache/solr/ltr/feature/TestFieldValueFeature.java

```diff
@@ -70,86 +78,92 @@ public void after() throws Exception {
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[0]/id=='1'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[1]/id=='8'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[2]/id=='6'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[3]/id=='7'");
+
+      query.add("rq", "{!ltr model="+field+"-model reRankDocs=4}");
+
+      assertJQ("/query" + query.toQueryString(), "/response/numFound/==4");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[0]/id=='8'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[1]/id=='7'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[2]/id=='6'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[3]/id=='1'");
+
+      query.setQuery("*:*");
+      query.remove("rows");
+      query.add("rows", "8");
+      query.remove("rq");
+      query.add("rq", "{!ltr model="+field+"-model reRankDocs=8}");
+
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[0]/id=='8'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[1]/id=='7'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[2]/id=='6'");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[3]/id=='5'");
+    }
   }
 
 
   @Test
   public void testIfADocumentDoesntHaveAFieldDefaultValueIsReturned() throws Exception {
-    SolrQuery query = new SolrQuery();
-    query.setQuery("id:42");
-    query.add("fl", "*, score");
-    query.add("rows", "4");
-
-    assertJQ("/query" + query.toQueryString(), "/response/numFound/==1");
-    assertJQ("/query" + query.toQueryString(), "/response/docs/[0]/id=='42'");
-    query = new SolrQuery();
-    query.setQuery("id:42");
-    query.add("rq", "{!ltr model=popularity-model reRankDocs=4}");
-    query.add("fl", "[fv]");
-    assertJQ("/query" + query.toQueryString(), "/response/numFound/==1");
-    assertJQ("/query" + query.toQueryString(),
-            "/response/docs/[0]/=={'[fv]':'"+FeatureLoggerTestUtils.toFeatureVector("popularity",Float.toString(FIELD_VALUE_FEATURE_DEFAULT_VAL))+"'}");
-
+    for (String field : FIELD_NAMES) {
+      SolrQuery query = new SolrQuery();
+      query.setQuery("id:42");
+      query.add("fl", "*, score");
+      query.add("rows", "4");
+
+      assertJQ("/query" + query.toQueryString(), "/response/numFound/==1");
+      assertJQ("/query" + query.toQueryString(), "/response/docs/[0]/id=='42'");
+      query = new SolrQuery();
+      query.setQuery("id:42");
+      query.add("rq", "{!ltr model="+field+"-model reRankDocs=4}");
+      query.add("fl", "[fv]");
+      assertJQ("/query" + query.toQueryString(), "/response/numFound/==1");
+      assertJQ("/query" + query.toQueryString(),
+              "/response/docs/[0]/=={'[fv]':'"+FeatureLoggerTestUtils.toFeatureVector(field,Float.toString(FIELD_VALUE_FEATURE_DEFAULT_VAL))+"'}");
```

Comment:
**This test fails here on the second and subsequent run of the loop.** I don't know yet if that is a test artifact or an actual code issue. The `testIfADocumentDoesntHaveAFieldASetDefaultValueIsReturned` test does not have the same behaviour since each loop turn uses a different `f(feature )store` (and its own model) whereas this test uses the test-wide models and its default feature store.

Reply:
I think the problem is `final private boolean extractAllFeatures;` in `LTRScoringQuery`. It is set to true and therefore all previous features are added to the response instead of just the model features.

This is the logic (line 210 in LTRScoringQuery):
```java
if (this.extractAllFeatures) {
    features = allFeatures;
} else{
    features =  modelFeatures;
}
```

I didn't find a way to set it to false with a query param.
So we would have two options:
- use this constructor `public LTRScoringQuery(LTRScoringModel ltrScoringModel, boolean extractAllFeatures)`
- adjust the xpath to something like `[ends-with(., '<fieldname>=0.0')]`

I am trying to adjust the xpath, because I would prefer that solution, but I was not able to find the right pattern so far.

Reply:
1/2 Thanks for investigating in detail here! I agree the explanation is with the `extractAllFeatures` boolean i.e. the response contains the values for all the features in the feature store. The puzzling thing then though is why the first run of the loop succeeds and only subsequent runs fail. Since all loop runs use the same feature store and surely each loop run should return the same feature values? ...